OpenRA.Mods.Common/Pathfinder/HierarchicalPathFinder.cs

```diff
@@ -718,24 +719,51 @@ static Grid GetGrid(CPos cellInGrid, Grid mapBounds)
 
 			pathFinderOverlay?.NewRecording(self, sources, target);
 
+			if (!world.Map.Contains(target))
+				return PathFinder.NoPath;
+
 			RebuildDirtyGrids();
 
 			var targetAbstractCell = AbstractCellForLocalCell(target);
 			if (targetAbstractCell == null)
 				return PathFinder.NoPath;
 
-			var sourcesWithReachableNodes = new List<CPos>(sources.Count);
-			var sourceEdges = new List<GraphEdge>(sources.Count);
+			// Unlike the target cell, the source cell is allowed to be an unreachable location.
+			// Instead, what matters is whether any cell adjacent to the source cell can be reached.
+			var sourcesWithReachableNodes = new List<(CPos Source, CPos AdjacentSource)>(sources.Count * CVec.Directions.Length);
+			var sourceEdges = new List<GraphEdge>(sources.Count * CVec.Directions.Length);
```

Comment:
I wonder, does it still make sense to initialize this with `* CVec.Directions.Length` considering that we most likely won't be falling back to checking the neighbors?

Reply:
It does not, removed.